backend/wrapped/serializers.py

```diff
@@ -0,0 +1,105 @@
+
+    class Meta:
+        model = IndividualStat
+        fields = ['key', 'value', 'semester']
+
+
+class GlobalStatSerializer(serializers.ModelSerializer):
+    key = serializers.StringRelatedField()
+
+    class Meta:
+        model = GlobalStat
+        fields = ['key', 'value', 'semester']
+
+
+class IndividualStatThroughSerializer(serializers.ModelSerializer):
+    individual_stat_value = serializers.SerializerMethodField()
+
+    class Meta:
+        model = IndividualStatThrough
+        fields = ['text_field_name', 'individual_stat_value']
+
+    def get_individual_stat_value(self, obj):
+        user = self.context.get('user')
+        semester = self.context.get('semester')
+
+        try:
+            individual_stat = IndividualStat.objects.get(
+                user=user,
+                key=obj.IndividualStatKey,
+                semester=semester
+            )
+            return individual_stat.value
+        except IndividualStat.DoesNotExist:
+            return None
+
+
+class GlobalStatThroughSerializer(serializers.ModelSerializer):
+    global_stat_value = serializers.SerializerMethodField()
+
+    class Meta:
+        model = GlobalStatThrough
+        fields = ['text_field_name', 'global_stat_value']
+
+    def get_global_stat_value(self, obj):
+        semester = self.context.get('semester')
+
+        try:
+            global_stat = GlobalStat.objects.get(
+                key=obj.GlobalStatKey.key,
+                semester=semester
+            )
+            return global_stat.value
+        except GlobalStat.DoesNotExist:
+            return None
+
+
+class PageSerializer(serializers.ModelSerializer):
+    combined_stats = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Page
+        fields = ['name', 'template_path', 'combined_stats']
+
+    def get_combined_stats(self, obj):
+        user = self.context.get('user')
+        semester = self.context.get('semester')
+
+        individual_throughs = IndividualStatThrough.objects.filter(Page=obj)
+        combined_stats = {}
+        for entry in individual_throughs:
+            individual_stat = IndividualStat.objects.filter(
+                user=user, key=entry.IndividualStatKey, semester=semester
+            ).first()
+            if individual_stat:
+                combined_stats[entry.text_field_name] = individual_stat.value
+
+        global_throughs = GlobalStatThrough.objects.filter(Page=obj)
+        for entry in global_throughs:
+            global_stat = GlobalStat.objects.filter(
+                key=entry.GlobalStatKey.key, semester=semester
+            ).first()
+            if global_stat:
+                combined_stats[entry.text_field_name] = global_stat.value
+
+        return combined_stats
+
+
+class SemesterSerializer(serializers.ModelSerializer):
+    pages = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Semester
+        fields = ['semester', 'pages']
+
+    def get_pages(self, obj):
```

Comment:
You are manually writing what django can already do if you just hook up pages directly with the serializer. (See sublet serializers for example). 
e.g.
```class SubletSerializerRead(serializers.ModelSerializer):
    amenities = serializers.PrimaryKeyRelatedField(
        many=True, queryset=Amenity.objects.all(), required=False
    )
    images = SubletImageURLSerializer(many=True, required=False)
```

Reply:
wait sorry i think for this you need to manually pass the semester obj so yours is correct. But its kinda verbose why not just: 

`return PageSerializer(obj.pages.all(), many=True, context={'semester': obj}).data`